src/main/java/htsjdk/samtools/SamReader.java

```diff
@@ -547,42 +547,27 @@ static AssertingIterator of(final CloseableIterator<SAMRecord> iterator) {
         }
 
         private final CloseableIterator<SAMRecord> wrappedIterator;
-        private SAMRecord previous = null;
-        private SAMRecordComparator comparator = null;
+        private SAMSortOrderChecker checker = null;
 
         public AssertingIterator(final CloseableIterator<SAMRecord> iterator) {
             wrappedIterator = iterator;
         }
 
         public SAMRecordIterator assertSorted(final SAMFileHeader.SortOrder sortOrder) {
-
-            if (sortOrder == null || sortOrder == SAMFileHeader.SortOrder.unsorted) {
-                comparator = null;
-                return this;
-            }
-
-            comparator = sortOrder.getComparatorInstance();
+            checker = new SAMSortOrderChecker(sortOrder);
```

Comment:
Can you add some unit tests for `assertSorted`? In particular, I'd like to see a test that takes in a bam that would fail the full comparator with its tie-breaking rules, but should pass the `assertSorted` check with its less-strict fileOrderCompare.